openquake/sep/liquefaction/liquefaction.py

```diff
@@ -113,18 +152,496 @@ def zhu_liquefaction_probability_general(
+    Megazonation of earthquake-induced soil liquefaction hazard in continental Europe.
+    Bulletin of Earthquake Engineering, 19(10), 4059–4082.
+    https://doi.org/10.1007/s10518-020-01008-6
+
+    :param pga:
+        Peak Ground Acceleration, measured in g
+    :param mag:
+        Magnitude of causative earthquake (moment or work scale)
+    :param cti:
+        Compound Topographic Index, a proxy for soil wetness.
+    :param vs30:
+        Shear-wave velocity averaged over the upper 30 m of the earth at the
+        site.
+
+    :returns:
+        prob_liq: Probability of liquefaction at the site.
+        out_class: Binary output 0 or 1, i.e., liquefaction nonoccurrence
+                   or liquefaction occurrence occurrence.
+    """
+    pga_scale = pga * _idriss_magnitude_weighting_factor(mag)
+    Xg = (
+        pgam_coeff * np.log(pga_scale)
+        + cti_coeff * cti
+        + vs30_coeff * np.log(vs30)
+        + intercept
+    )
+    prob_liq = sigmoid(Xg)
+    out_class = np.where(prob_liq > 0.57, 1, 0)
+    return prob_liq, out_class
+
+
+def todorovic_silva_2022_nonparametric_general(
+    pgv: Union[float, np.ndarray],
+    vs30: Union[float, np.ndarray],
+    dw: Union[float, np.ndarray],
+    wtd: Union[float, np.ndarray],
+    precip: Union[float, np.ndarray],
+    ) -> Union[float, np.ndarray]:
+    """
+    Returns the binary class output (i.e, 0 or 1) which indicates liquefaction
+    nonoccurrence or liquefaction occurrence as per Todorovic and Silva (2022). 
+    In addition, it returns the probability of belonging to a class 1, or liq-
+    uefaction occurrence. The implemented model includes modifications from the 
+    version published in Todorovic and Silva (2022).
+
+    Reference: Todorovic, L., Silva, V. (2022). 
+    A liquefaction occurrence model for regional analysis. 
+    Soil Dynamics and Earthquake Engineering, 161, 1–12. 
+    https://doi.org/10.1016/j.soildyn.2022.107430
+
+    :param pgv:
+        Peak Ground Velocity, measured in cm/s
+    :param vs30:
+        Shear-wave velocity averaged over the upper 30 m of the earth at the
+        site, measured in m/s
+    :param dw:
+        Distance to the nearest water body, measured in km
+    :param wtd:
+        Global water table depth, measured in m
+    :param precip:
+        Mean annual precipitation, measured in mm
+
+    :returns:
+        out_class: ndarray of len(sitemodel)  
+            Returns binary output 0 or 1, i.e., liquefaction nonoccurrence or
+            liquefaction occurrence occurrence.
+        out_prob: ndarray of len(sitemodel)
+            Returns probability of belonging to class 1.
+        
+    """
+    strain_proxy = pgv / (CM_PER_M * vs30)
+    dict = {
+        'strain_proxy': strain_proxy,
+        'dw': dw,
+        'wtd': wtd,
+        'precip': precip
+    }
+    df = pd.DataFrame(dict)
+    model_file = 'data/todorovic_silva_2022/todorovic_silva_2022.pkl.gz'
```

Comment:
Committing a pickled file is a REALLY BAD idea. What is inside there and how big is it? Could we store it in an alternative format like JSON or so? This is enough to REJECT this pull request. The reason: pickled files becomes immediately unreadable as soon as the code changes (our code or third party code) and cannot by maintained. I have been there and seen that. Pickle files are okay only as temporary files, not as a persistent storage.

Reply:
The pkl file contains the trained ML model, and we’re simply following the instructions [here]( https://scikit-learn.org/stable/model_persistence.html). Will look further into alternatives.

Reply:
Can't we commit the inputs from which the trained ML is generated? I understand that then every user will have to rebuild it the first time he use the secondary perils module, but perhaps it is acceptable?

Reply:
My concern is not about security (yes, there is a huge problem of security, but that is a problem of @vot4anto), my problem is about reproducibility. As soon as anything change (any one of the 1000 dependencies of the engine or of scikit-learn) the pickle file may become unreadable. As they say in the docs:
```
While models saved using one version of scikit-learn might load in other versions, this is entirely unsupported and inadvisable. It should also be kept in mind that operations performed on such data could give different and unexpected results.
```
This is huge for me, since I know that our users make a mess with dependencies already.

Reply:
There are a number of challenges to overcome here:
1. pickle is fundamentally insecure and dependent on Python version while the model file will also be tied to the scikit-learn version (and recursively all of scikit's dependencies...).   We should investigate use of the alternative formats ONNX and PMML (see [https://scikit-learn.org/stable/model_persistence.html](https://scikit-learn.org/stable/model_persistence.html) )
3. Storing large binary files in the oq-engine repository implies that _all_ oq-engine git users (even those who do not use or care about ML) will have to install and configure git-lfs.   This is not impossible but is extremely disruptive and is in any case not without problems for e.g. Windows users.   
4. If we need to store many (versions of) binary files in the engine repository this will further increase the size of the repository, increase the time required to clone/download and slow down our automated tests. 
5. If instead we store the model files outside of the engine repository, we need our own (possibly non-git) repository of model files with versions and branches that are (manually) kept in sync with the git repo.  This obviously requires additional testing.
6. Ensuring that only GEM-authored models are used by oq-engine (as suggested by @VSilva ) further complicates matters with digital signatures etc. would prevent others from being able to develop their own models and is in any case not necessary if we can use a secure, non-pickle format.   It is still necessary to detect and prevent loading corrupt model files (this can be achieved with a SHA-1 or similar hash control).
7. We must be able to regenerate the model files automatically in order to e.g. upgrade to newer versions of scikit-learn.

Let's schedule a meeting to talk about this.

Reply:
More on this (point by point):

1. It is nice to see that we might have some alternatives to pickle, though both of these formats seems rather experimental (according to what is written on the link). Whatever format we choose, it is important to make sure it can be readily read/written by scikit-learn to streamline the process.
2. I do not think this will be a problem at all. The most complex ML model we have right now is 7MB. The most complex models we created for vulnerability modelling are 30 KB each one (so very small). We could also explore the gitignore option if we want users to have the option to exclude ML models.
3. Again, the files we have produced now are rather small.
4. The manual part of this process seems like it will bring pain and a lot of work in the future.
5. This is valid if an alternative format is available (and I wish there is, but at least according to the brief reading I did online, it does not seem (yet) to be the case). Providing the training datasets is common practice in some of the most common ML platforms, so we would not be asking nothing new to the users if we need (and should) do the training on our side before committing the files. I do not think this would prevent users from providing their own models at all. 
6. I doubt the process will ever be fully automatic. With newer versions of scikit-learn even if we use the same training datasets and hyperparametrs, there is a chance that the performance will be different (and the hyperparametrs might have to be adjusted). We need to check this so the process will be semi-automatic at most.

Let's talk soon.